pkg/operator/ceph/object/health.go

```diff
@@ -0,0 +1,269 @@
+)
+
+// rgwStatusChecker aggregates the mon/cluster info needed to check the health of the monitors
+type rgwStatusChecker struct {
+	context        *clusterd.Context
+	objContext     *Context
+	interval       time.Duration
+	serviceIP      string
+	port           string
+	client         client.Client
+	namespacedName types.NamespacedName
+}
+
+// newRgwStatusChecker creates a new HealthChecker object
+func newRgwStatusChecker(context *clusterd.Context, objContext *Context, serviceIP, port string, client client.Client, namespacedName types.NamespacedName) *rgwStatusChecker {
+	c := &rgwStatusChecker{
+		context:        context,
+		objContext:     objContext,
+		interval:       defaultStatusCheckInterval,
+		serviceIP:      serviceIP,
+		port:           port,
+		namespacedName: namespacedName,
+		client:         client,
+	}
+
+	// allow overriding the check interval with an env var on the operator
+	checkInterval := os.Getenv("ROOK_CEPH_OBJECT_STATUS_CHECK_INTERVAL")
+	if checkInterval != "" {
+		if duration, err := time.ParseDuration(checkInterval); err == nil {
+			logger.Infof("ceph rgw status check interval is %q", checkInterval)
+			c.interval = duration
+		}
+	}
+
+	return c
+}
+
+// checkCephStatus periodically checks the health of the cluster
+func (c *rgwStatusChecker) checkObjectStore(stopCh chan struct{}) {
+
+	for {
+		select {
+		case <-stopCh:
+			// purge bucket and s3 user
+			// Needed for external mode where in converged everything goes away with the CR deletion
+			c.cleanupHealthCheck()
+			logger.Infof("stopping monitoring of rgw endpoints")
+			return
+
+		case <-time.After(defaultStatusCheckInterval):
+			logger.Debug("checking rgw health")
+			err := c.checkObjectStoreHealth()
+			if err != nil {
+				logger.Warningf("failed to check rgw health. %v", err)
+			}
+		}
+	}
+}
+
+func (c *rgwStatusChecker) checkObjectStoreHealth() error {
+	/*
+		0. purge everything by default
+		1. create an S3 user
+		2. always use the same user
+		3. if already exists just re-hydrate the s3 credentials
+		4. create a bucket with that user or use the existing one (always use the same bucket)
+		5. create a check file
+		6. get the MD5 of the file
+		7. PUT the file
+		8. GET the file
+		9. compare MD5
+		10. delete object on bucket
+		11. update CR health status check
+
+		Always keep the bucket and the user for the healthcheck, just do PUT and GET because bucket creation is expensive
+	*/
+
+	var s3AccessKey string
+	var s3SecretKey string
+	s3endpoint := fmt.Sprintf("%s:%s", BuildDomainName(c.objContext.Name, c.namespacedName.Namespace), c.port)
+
+	// Generate unique user and bucket name
+	bucket := genUniqueBucketName(c.objContext.UID)
+	userConfig := c.genUserConfig()
+
+	// Create S3 user
+	logger.Debugf("creating s3 user object %q", userConfig.UserID)
+	user, rgwerr, err := CreateUser(c.objContext, userConfig)
+	if err != nil {
+		if rgwerr == ErrorCodeFileExists {
+			objectUser, _, err := GetUser(c.objContext, userConfig.UserID)
```

Comment:
If you set the same user object, you wouldn't need lines 138-139, then lines 148-149 would also apply in this case if you removed them from the `else` statement.
```suggestion
			user, _, err = GetUser(c.objContext, userConfig.UserID)
```